### ISSUE TITLE
Rule syntax error

### ISSUE THREAD

**Issue by UPI05** [coreruleset/coreruleset#3610]
Syntax error on these rules:
- REQUEST-920-PROTOCOL-ENFORCEMENT.conf
- REQUEST-922-MULTIPART-ATTACK.conf
- REQUEST-934-APPLICATION-ATTACK-GENERIC.conf

**Comment by superlgn** [comment]
I'm seeing the same with Apache 2.4.57-2 on Debian 12:

(apache2ctl configtest)
AH00526: Syntax error on line 1256 of crs/rules/REQUEST-920-PROTOCOL-ENFORCEMENT.conf:
Error creating rule: Error compiling pattern (offset 167): (null)
AH00526: Syntax error on line 76 of crs/rules/REQUEST-922-MULTIPART-ATTACK.conf:
Error creating rule: Error compiling pattern (offset 167): (null)
AH00526: Syntax error on line 234 of crs/rules/REQUEST-934-APPLICATION-ATTACK-GENERIC.conf:
Error creating rule: Error compiling pattern (offset 172): (null)

These rules are all using the charset regex. If I replace that regex with the old one from 4.0/dev or comment the rules out then Apache is able to load everything.

**Comment by airween** [comment]
@UPI05, @superlgn - guys, thank you for sending your report.

Please take a look to our issue [template](https://github.com/coreruleset/coreruleset/issues/new?assignees=&labels=%3Abug%3A+bug&projects=&template=03_bug-report.md&title=), and fill it correctly.

It's very important us to know which version of mod_security2 do you use.

Personally, I use Apache2 + mod_security2 (2.9.7) on Debian 12, both with CRS 3.3.5 and 4.0.0, without any error. CRS CI/CD tests also use that (that's our reference system).

I assume you guys use some 3rd-party repository which uses some strange/badly/incompatible built regex engine.

**Comment by superlgn** [comment]
* CRS version (e.g., v3.2.0): main (new clone as of today)
* Paranoia level setting: PL1
* ModSecurity version (e.g., 2.9.3): 2.9.7-1+b1
* Web Server and version (e.g., apache 2.4.41): 2.4.57-2
* Operating System and version: Debian 12.5 (up to date as of today)

I've got a handful of 3rd party repos setup for other packages like virtualbox, firefox, anydesk, slack, docker, and syncthing, but I don't believe any of them would provide Apache, ModSecurity, or anything PCRE.

# apt-show-versions|egrep '(apache2|mod-sec|pcre2-8)'
apache2:amd64/bookworm 2.4.57-2 uptodate
apache2-bin:amd64/bookworm 2.4.57-2 uptodate
apache2-data:all/bookworm 2.4.57-2 uptodate
apache2-utils:amd64/bookworm 2.4.57-2 uptodate
libapache2-mod-security2:amd64/bookworm 2.9.7-1+b1 uptodate
libpcre2-8-0:amd64/bookworm 10.42-1 uptodate

These all appear to be the correct versions.

# ldd /usr/lib/apache2/modules/mod_security2.so|grep pcre
	libpcre2-8.so.0 => /lib/x86_64-linux-gnu/libpcre2-8.so.0 (0x00007fdecb950000)

(same for apache2)

# dpkg -L libpcre2-8-0|grep libpcre2-8.so
/usr/lib/x86_64-linux-gnu/libpcre2-8.so.0.11.2
/usr/lib/x86_64-linux-gnu/libpcre2-8.so.0

# locate libpcre2-8.so
/usr/lib/x86_64-linux-gnu/libpcre2-8.so
/usr/lib/x86_64-linux-gnu/libpcre2-8.so.0
/usr/lib/x86_64-linux-gnu/libpcre2-8.so.0.11.2

/lib is a symlink to /usr/lib, no other libpcre2-8.so* anywhere else.

No output from "dpkg -V libapache2-mod-security2 libpcre2-8-0".

I even reinstalled both to make sure they came from debian.org:

# aptitude reinstall libapache2-mod-security2 libpcre2-8-0
The following packages will be REINSTALLED:
  libapache2-mod-security2 libpcre2-8-0 
0 packages upgraded, 0 newly installed, 2 reinstalled, 0 to remove and 1 not upgraded.
Need to get 520 kB of archives. After unpacking 0 B will be used.
Get: 1 http://deb.debian.org/debian bookworm/main amd64 libpcre2-8-0 amd64 10.42-1 [261 kB]
Get: 2 http://deb.debian.org/debian bookworm/main amd64 libapache2-mod-security2 amd64 2.9.7-1+b1 [259 kB]
Fetched 520 kB in 0s (3780 kB/s)                   
(Reading database ... 325398 files and directories currently installed.)
Preparing to unpack .../libpcre2-8-0_10.42-1_amd64.deb ...
Unpacking libpcre2-8-0:amd64 (10.42-1) over (10.42-1) ...
Setting up libpcre2-8-0:amd64 (10.42-1) ...
(Reading database ... 325398 files and directories currently installed.)
Preparing to unpack .../libapache2-mod-security2_2.9.7-1+b1_amd64.deb ...
Unpacking libapache2-mod-security2 (2.9.7-1+b1) over (2.9.7-1+b1) ...
Setting up libapache2-mod-security2 (2.9.7-1+b1) ...
...

Added bookworm-backports to my system just in case something was waiting there. (There wasn't.)

My Debian 12 desktop at home has a nearly identical configuration as work, both fail.

But FWIW I tested the rules on one of each of my Debian 10 and 11 servers and they compile just fine.

So I must be missing something here. If I manage to figure it out I'll follow up.

**Comment by airween** [comment]
Hi @superlgn,

thanks for detailed report!

After the steps above (in your last comment), is the problem still exist? I mean the engine complains about syntax errors at rules?

**Comment by airween** [comment]
Could you take a try with this repository?

https://modsecurity.digitalwave.hu

I just finished a fresh Debian 12 install now, with CRS4 (not from package - but the mentioned rules are the same), and everything works as well.

**Comment by airween** [comment]
hi @superlgn,

first of all, many-many thanks for your hard work.

Actually I can't say anything, I totally have no idea what's wrong with that setup. I think I'm going to check the source code of mod_security2 module for Apache, where is this error message...

Thanks again!

**Comment by superlgn** [comment]
I thought maybe it had something to do with SecPcreMatchLimit/Recursion but I've tried all different values there. I've also taken the regex from rule 920600 and dumped it into an old pcretest binary linked against pcre3 and it doesn't complain (not sure what version of Debian I originally snagged that from but it's dated 2010). It also compiles with preg_match() in php-cli-8.2.7 which is linked against libpcre2-8.so.0.

I wondered if this might be an issue with Apache 2.4.57-2's config parser getting confused by something in the regex, causing it to send a truncated or otherwise mangled value to modsecurity, however unlikely that may be. So I downgraded apache2, -bin, -data, -doc, -utils (and libssl1.1, libldap-2.4-2, libpcre3 so I could install apache2*), and libpcre2-8-0 again, but the syntax error still persisted.

Finally I downgraded libapache2-mod-security2 to 2.9.3-3+deb11u2 from oldstable and that did it. I'm now able to successfully run apache2ctl configtest. I put libapache2-mod-security2 on hold and brought everything else back up to date, and it's still good.

Must be something with apache modsecurity module greater than 2.9.3-3+deb11u2.

The instructions on the Digitalwave page are heavy on nginx. Can I assume that's what you're testing with, and that's how you've managed to go unscathed?

Regardless, this appears like it may not be crs issue so this conversation may be happening at the wrong place.

**Comment by airween** [comment]
It's really weird.

Now I removed Digitalwave's repository from the new Debian 12 system completely, and reinstalled mod_security2.

This is what I see in log:
```
[Wed Mar 13 22:08:15.737012 2024] [security2:notice] [pid 27398] ModSecurity for Apache/2.9.7 (http://www.modsecurity.org/) configured.
[Wed Mar 13 22:08:15.737124 2024] [security2:notice] [pid 27398] ModSecurity: APR compiled version="1.7.2"; loaded version="1.7.2"
[Wed Mar 13 22:08:15.737133 2024] [security2:notice] [pid 27398] ModSecurity: PCRE2 compiled version="10.42 "; loaded version="10.42 2022-12-11"
[Wed Mar 13 22:08:15.737139 2024] [security2:notice] [pid 27398] ModSecurity: LUA compiled version="Lua 5.1"
[Wed Mar 13 22:08:15.737144 2024] [security2:notice] [pid 27398] ModSecurity: YAJL compiled version="2.1.0"
[Wed Mar 13 22:08:15.737150 2024] [security2:notice] [pid 27398] ModSecurity: LIBXML compiled version="2.9.14"
[Wed Mar 13 22:08:15.737155 2024] [security2:notice] [pid 27398] ModSecurity: Status engine is currently disabled, enable it by set SecStatusEngine to On.
[Wed Mar 13 22:08:15.951682 2024] [mpm_prefork:notice] [pid 27399] AH00163: Apache/2.4.57 (Debian) configured -- resuming normal operations
```

with CRS4. So I use all of the mentioned rules above...

Are you on AMD64, right?

> Finally I downgraded libapache2-mod-security2 to 2.9.3-3+deb11u2 from oldstable and that did it.

The difference between the two packages is not just the mod_security2 version, but in Debian 11 we made the package with libpcre3, and in Debian 12 with libpcre2 (the new regex library). So I still suspect some pcre issue.

If you think, you can download the source of the package, remove the `--with-pcre2` flag from [here](https://salsa.debian.org/modsecurity-packaging-team/modsecurity-apache/-/blob/debian/2.9.7-1/debian/rules?ref_type=tags#L19).

Also you need to align the dependencies [here](https://salsa.debian.org/modsecurity-packaging-team/modsecurity-apache/-/blob/debian/2.9.7-1/debian/control?ref_type=tags#L5).

**Comment by superlgn** [comment]
Yep, AMD64 here.

Alright, so I rebuilt libapache2-mod-security2 2.9.7-1+b1 without the --with-pcre2 flags and mod_security2.so was indeed built with and linked against the older pcre3, and now apache2ctl configtest works. For the sake of completeness I also rebuilt it a second time using the now normal pcre2 flags per debian/rules, and it built as expected and configtest once again fails.

If Debian 12's pcre2 build of php-cli-8.2.7 is able to compile the regex with pcre_match() then perhaps this isn't an issue with the generated regex in the crs ruleset but rather the pcre2 library usage in the module.

I tried building it one more time using --with-pcre2 but without --enable-pcre-jit --enable-pcre-study and configtest still fails. Removing the PCRE2_ZERO_TERMINATED flag works (configtest: ok), but this should be a null terminated string. All other usage of the pcre2_compile() function I've seen has this flag. I'm admittedly unfamiliar with the pcre library. My only experience with regular expressions in C is POSIX regcomp() type stuff.

**Comment by airween** [comment]
@superlgn,

many-many thanks, this is an awesome report.

pcre2 packages are from the official Debian mirror, right?

Btw in the past I tried Digitalwave's mod_security2 packages both from Debian's official pcre2 packages and packages from Sury's repository - but as I wrote I've never met this behavior before.

Actually I have no idea, how can we fix this issue. Let me think about this.

**Comment by superlgn** [comment]
Yep, official pcre2 from Debian. I've got all different hardware configurations on the 3 affected Debian 12 systems. My test VM is running on xcp-ng. My work desktop is a 2009 Intel iMac running Debian bare metal. At home I've got a custom PC with an ASUS motherboard an a Xeon processor. Memory ranges from 1-64GB. The VM has no tweaks whatsoever, it's a straight up plain jane 12.5 minimalish install. My work and home systems are similar but not identical in software packages or configuration. They're both relatively new installs, upgraded from Debian 11. I have no why we're seeing this but you aren't. There's got to be an explanation.

**Comment by superlgn** [comment]
Hey, so I wrote a little pcre2.c test program to try and compile the problematic regex:

```
/*
 for i in 8 16 32
 do
   gcc -DPCRE2_CODE_UNIT_WIDTH=$i pcre2.c -o pcre2-$i `pcre2-config --libs$i`
 done
*/

#include <stdio.h>
#include <pcre2.h>

int main(int argc, char *argv[]) {
  int i;

  if (argc < 2) {
    printf("Usage: %s <regex1> <regex2> ...\n", argv[0]);
    exit(1);
  }

  for (i = 1; i < argc; i++) {
    int error_number = 0;
    PCRE2_SIZE error_offset = 0;

    pcre2_code *re;

    re = pcre2_compile((PCRE2_SPTR)argv[i],
                       PCRE2_ZERO_TERMINATED,
                       0,
                       &error_number,
                       &error_offset,
                       NULL);

    if (re) {
      printf("%s compiled successfully\n", argv[i]);
    } else {
      PCRE2_UCHAR error[128];

      pcre2_get_error_message(error_number, error, sizeof(error));

      printf("%s failed to compile: %s (errno %d) at offset %d\n",
             argv[i], error, error_number, error_offset);
    }
  }

  exit(0);
}
```

and the charset regex for 920600 won't compile with PCRE2_CODE_UNIT_WIDTH=8 but does with 16 or 32.

When I try it with 8 I get: ... failed to compile: invalid range in character class: errno 150 at offset 171. offset 171 is near the hyphen in (?:[^\s\v-\"\(\). If I escape that hyphen (\-) then the error moves further down the line. This is a pretty gnarly regular expression. I like to think I'm pretty good with regular expressions but this is rough. I don't know exactly what the intention is here, if this actually supposed to be a range or just a list of characters to not match on, eg [^\s\v\-\"\(\)].

I ran across https://php.watch/versions/7.3/pcre2 which says that pcre2 is less forgiving when a bare hyphen is involved in the character grouping like this. Same error as the above.

**Comment by airween** [comment]
> Yep, official pcre2 from Debian. I've got all different hardware configurations on the 3 affected Debian 12 systems. My test VM is running on xcp-ng. My work desktop is a 2009 Intel iMac running Debian bare metal. At home I've got a custom PC with an ASUS motherboard an a Xeon processor. Memory ranges from 1-64GB. The VM has no tweaks whatsoever, it's a straight up plain jane 12.5 minimalish install. My work and home systems are similar but not identical in software packages or configuration. They're both relatively new installs, upgraded from Debian 11. I have no why we're seeing this but you aren't. There's got to be an explanation.

Could you show your `/proc/cpuinfo` outputs from those machines?

**Comment by superlgn** [comment]
Here's my xcp-ng VM:
```
processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 23
model name	: Intel(R) Xeon(R) CPU           E5420  @ 2.50GHz
stepping	: 6
microcode	: 0x60f
cpu MHz		: 2500.108
cache size	: 6144 KB
physical id	: 0
siblings	: 1
core id		: 0
cpu cores	: 1
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 10
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush acpi mmx fxsr sse sse2 syscall nx lm constant_tsc rep_good nopl cpuid tsc_known_freq pni ssse3 cx16 sse4_1 hypervisor lahf_lm cpuid_fault pti
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs itlb_multihit mmio_unknown
bogomips	: 5000.22
clflush size	: 64
cache_alignment	: 64
address sizes	: 38 bits physical, 48 bits virtual
power management:
```

My work iMac (the first core):
```
processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 30
model name	: Intel(R) Core(TM) i5 CPU         750  @ 2.67GHz
stepping	: 5
microcode	: 0xa
cpu MHz		: 1196.959
cache size	: 8192 KB
physical id	: 0
siblings	: 4
core id		: 0
cpu cores	: 4
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 11
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ht tm pbe syscall nx rdtscp lm constant_tsc arch_perfmon pebs bts nopl xtopology nonstop_tsc cpuid aperfmperf pni dtes64 monitor ds_cpl vmx smx est tm2 ssse3 cx16 xtpr pdcm sse4_1 sse4_2 popcnt lahf_lm ssbd ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid dtherm ida flush_l1d
vmx flags	: vnmi preemption_timer invvpid ept_x_only flexpriority tsc_offset vtpr mtf vapic ept vpid
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs itlb_multihit mmio_unknown
bogomips	: 5319.70
clflush size	: 64
cache_alignment	: 64
address sizes	: 36 bits physical, 48 bits virtual
power management:
```

And my home machine:
```
processor	: 0
vendor_id	: GenuineIntel
cpu family	: 6
model		: 45
model name	: Intel(R) Xeon(R) CPU E5-2670 0 @ 2.60GHz
stepping	: 6
microcode	: 0x621
cpu MHz		: 1857.482
cache size	: 20480 KB
physical id	: 0
siblings	: 16
core id		: 0
cpu cores	: 8
apicid		: 0
initial apicid	: 0
fpu		: yes
fpu_exception	: yes
cpuid level	: 13
wp		: yes
flags		: fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic popcnt tsc_deadline_timer aes xsave avx lahf_lm epb ssbd ibrs ibpb stibp tpr_shadow vnmi flexpriority ept vpid xsaveopt dtherm arat pln pts md_clear flush_l1d
vmx flags	: vnmi preemption_timer invvpid ept_x_only ept_1gb flexpriority tsc_offset vtpr mtf vapic ept vpid unrestricted_guest ple
bugs		: cpu_meltdown spectre_v1 spectre_v2 spec_store_bypass l1tf mds swapgs itlb_multihit mmio_unknown
bogomips	: 5203.22
clflush size	: 64
cache_alignment	: 64
address sizes	: 46 bits physical, 48 bits virtual
power management:
```

They're all older CPUs.

**Comment by airween** [comment]
Hi,

thanks for the code, this is what I planned to write, so now you save some time for me :smiley:

**Comment by airween** [comment]
Hi @superlgn,

thanks again for your example. Unfortunately I wasn't able to use that so I created a new one.

It's a gist, you (or anyone) can access it [here](https://gist.github.com/airween/5c08626531c97dc63c2f57d9e4bf068e).

Please take a look to comments between source code lines.

There is an important thing: mod_security2 uses Apache's parser to parse the config files. When Apache parses a config file, it calls a function which removes the extra slashes - if the token is quoted.

Example:

this will be stripped:
```
"\\string"
```
and the result is this:
```
"\string"
```

but not this one:
```
string
```
which will be keep as it is, because it's not quoted.

[Here](https://github.com/digitalwave/msc_retest?tab=readme-ov-file#parsing) you can read more info.

And - drumbeat - I was able to emulate the behavior that you explained :(

```
$ ./pcre2test 920600.txt 
Original pattern: '^(?:(?:\*|[^!\"\(\),/:-\?\[-\]\{\}]+)/(?:\*|[^!\"\(\),/:-\?\[-\]\{\}]+)|\*)(?:[\s\v]*;[\s\v]*(?:charset[\s\v]*=[\s\v]*\"?(?:iso-8859-15?|utf-8|windows-1252)\b\"?|(?:[^\s\v-\"\(\),/:-\?\[-\]c\{\}]|c(?:[^!\"\(\),/:-\?\[-\]h\{\}]|h(?:[^!\"\(\),/:-\?\[-\]a\{\}]|a(?:[^!\"\(\),/:-\?\[-\]r\{\}]|r(?:[^!\"\(\),/:-\?\[-\]s\{\}]|s(?:[^!\"\(\),/:-\?\[-\]e\{\}]|e[^!\"\(\),/:-\?\[-\]t\{\}]))))))[^!\"\(\),/:-\?\[-\]\{\}]*[\s\v]*=[\s\v]*[^!\(\),/:-\?\[-\]\{\}]+);?)*(?:[\s\v]*,[\s\v]*(?:(?:\*|[^!\"\(\),/:-\?\[-\]\{\}]+)/(?:\*|[^!\"\(\),/:-\?\[-\]\{\}]+)|\*)(?:[\s\v]*;[\s\v]*(?:charset[\s\v]*=[\s\v]*\"?(?:iso-8859-15?|utf-8|windows-1252)\b\"?|(?:[^\s\v-\"\(\),/:-\?\[-\]c\{\}]|c(?:[^!\"\(\),/:-\?\[-\]h\{\}]|h(?:[^!\"\(\),/:-\?\[-\]a\{\}]|a(?:[^!\"\(\),/:-\?\[-\]r\{\}]|r(?:[^!\"\(\),/:-\?\[-\]s\{\}]|s(?:[^!\"\(\),/:-\?\[-\]e\{\}]|e[^!\"\(\),/:-\?\[-\]t\{\}]))))))[^!\"\(\),/:-\?\[-\]\{\}]*[\s\v]*=[\s\v]*[^!\(\),/:-\?\[-\]\{\}]+);?)*)*$'
Stripped pattern: '^(?:(?:\*|[^!"\(\),/:-\?\[-\]\{\}]+)/(?:\*|[^!"\(\),/:-\?\[-\]\{\}]+)|\*)(?:[\s\v]*;[\s\v]*(?:charset[\s\v]*=[\s\v]*"?(?:iso-8859-15?|utf-8|windows-1252)\b"?|(?:[^\s\v-"\(\),/:-\?\[-\]c\{\}]|c(?:[^!"\(\),/:-\?\[-\]h\{\}]|h(?:[^!"\(\),/:-\?\[-\]a\{\}]|a(?:[^!"\(\),/:-\?\[-\]r\{\}]|r(?:[^!"\(\),/:-\?\[-\]s\{\}]|s(?:[^!"\(\),/:-\?\[-\]e\{\}]|e[^!"\(\),/:-\?\[-\]t\{\}]))))))[^!"\(\),/:-\?\[-\]\{\}]*[\s\v]*=[\s\v]*[^!\(\),/:-\?\[-\]\{\}]+);?)*(?:[\s\v]*,[\s\v]*(?:(?:\*|[^!"\(\),/:-\?\[-\]\{\}]+)/(?:\*|[^!"\(\),/:-\?\[-\]\{\}]+)|\*)(?:[\s\v]*;[\s\v]*(?:charset[\s\v]*=[\s\v]*"?(?:iso-8859-15?|utf-8|windows-1252)\b"?|(?:[^\s\v-"\(\),/:-\?\[-\]c\{\}]|c(?:[^!"\(\),/:-\?\[-\]h\{\}]|h(?:[^!"\(\),/:-\?\[-\]a\{\}]|a(?:[^!"\(\),/:-\?\[-\]r\{\}]|r(?:[^!"\(\),/:-\?\[-\]s\{\}]|s(?:[^!"\(\),/:-\?\[-\]e\{\}]|e[^!"\(\),/:-\?\[-\]t\{\}]))))))[^!"\(\),/:-\?\[-\]\{\}]*[\s\v]*=[\s\v]*[^!\(\),/:-\?\[-\]\{\}]+);?)*)*$'
Regex compilation failed.
Error compiling pattern (offset 167): (null)
```
Actually I don't have any explanation why I'm able to reproduce this, and why does it work in real environment.

Any ideas are welcome. (Or source checking, may be I was wrong in some changes.)

**Comment by airween** [comment]
I updated the gist, now I get same as you:

```
PCRE2 compiled version="10.42"; loaded version="10.42 2022-12-11"
Original pattern: '^(?:(?:\*|[^!\"\(\),/:-\?\[-\]\{\}]+)/(?:\*|[^!\"\(\),/:-\?\[-\]\{\}]+)|\*)(?:[\s\v]*;[\s\v]*(?:charset[\s\v]*=[\s\v]*\"?(?:iso-8859-15?|utf-8|windows-1252)\b\"?|(?:[^\s\v-\"\(\),/:-\?\[-\]c\{\}]|c(?:[^!\"\(\),/:-\?\[-\]h\{\}]|h(?:[^!\"\(\),/:-\?\[-\]a\{\}]|a(?:[^!\"\(\),/:-\?\[-\]r\{\}]|r(?:[^!\"\(\),/:-\?\[-\]s\{\}]|s(?:[^!\"\(\),/:-\?\[-\]e\{\}]|e[^!\"\(\),/:-\?\[-\]t\{\}]))))))[^!\"\(\),/:-\?\[-\]\{\}]*[\s\v]*=[\s\v]*[^!\(\),/:-\?\[-\]\{\}]+);?)*(?:[\s\v]*,[\s\v]*(?:(?:\*|[^!\"\(\),/:-\?\[-\]\{\}]+)/(?:\*|[^!\"\(\),/:-\?\[-\]\{\}]+)|\*)(?:[\s\v]*;[\s\v]*(?:charset[\s\v]*=[\s\v]*\"?(?:iso-8859-15?|utf-8|windows-1252)\b\"?|(?:[^\s\v-\"\(\),/:-\?\[-\]c\{\}]|c(?:[^!\"\(\),/:-\?\[-\]h\{\}]|h(?:[^!\"\(\),/:-\?\[-\]a\{\}]|a(?:[^!\"\(\),/:-\?\[-\]r\{\}]|r(?:[^!\"\(\),/:-\?\[-\]s\{\}]|s(?:[^!\"\(\),/:-\?\[-\]e\{\}]|e[^!\"\(\),/:-\?\[-\]t\{\}]))))))[^!\"\(\),/:-\?\[-\]\{\}]*[\s\v]*=[\s\v]*[^!\(\),/:-\?\[-\]\{\}]+);?)*)*$'
Stripped pattern: '^(?:(?:\*|[^!"\(\),/:-\?\[-\]\{\}]+)/(?:\*|[^!"\(\),/:-\?\[-\]\{\}]+)|\*)(?:[\s\v]*;[\s\v]*(?:charset[\s\v]*=[\s\v]*"?(?:iso-8859-15?|utf-8|windows-1252)\b"?|(?:[^\s\v-"\(\),/:-\?\[-\]c\{\}]|c(?:[^!"\(\),/:-\?\[-\]h\{\}]|h(?:[^!"\(\),/:-\?\[-\]a\{\}]|a(?:[^!"\(\),/:-\?\[-\]r\{\}]|r(?:[^!"\(\),/:-\?\[-\]s\{\}]|s(?:[^!"\(\),/:-\?\[-\]e\{\}]|e[^!"\(\),/:-\?\[-\]t\{\}]))))))[^!"\(\),/:-\?\[-\]\{\}]*[\s\v]*=[\s\v]*[^!\(\),/:-\?\[-\]\{\}]+);?)*(?:[\s\v]*,[\s\v]*(?:(?:\*|[^!"\(\),/:-\?\[-\]\{\}]+)/(?:\*|[^!"\(\),/:-\?\[-\]\{\}]+)|\*)(?:[\s\v]*;[\s\v]*(?:charset[\s\v]*=[\s\v]*"?(?:iso-8859-15?|utf-8|windows-1252)\b"?|(?:[^\s\v-"\(\),/:-\?\[-\]c\{\}]|c(?:[^!"\(\),/:-\?\[-\]h\{\}]|h(?:[^!"\(\),/:-\?\[-\]a\{\}]|a(?:[^!"\(\),/:-\?\[-\]r\{\}]|r(?:[^!"\(\),/:-\?\[-\]s\{\}]|s(?:[^!"\(\),/:-\?\[-\]e\{\}]|e[^!"\(\),/:-\?\[-\]t\{\}]))))))[^!"\(\),/:-\?\[-\]\{\}]*[\s\v]*=[\s\v]*[^!\(\),/:-\?\[-\]\{\}]+);?)*)*$'
Regex compilation failed. error_number: 150
Error: invalid range in character class
Error compiling pattern (offset 167): (null)
|(?:[^\s\v-"\(\),/:
          ^        
```

**Comment by airween** [comment]
@superlgn, @UPI05 - would you try to check your WAF with [this](https://github.com/coreruleset/coreruleset/archive/refs/tags/v4.0.0.tar.gz) CRS source?

Try do get it:
```
wget https://github.com/coreruleset/coreruleset/archive/refs/tags/v4.0.0.tar.gz
```

and replace the existing one with that. Let me know your results.

**Comment by superlgn** [comment]
4.0.0 is fine for me. It's using the old regex, which is what I currently have copied into my production modsecurity ruleset, which incorporates crs main as well as some custom rules. The problem started somewhere in 4.0.1, approximately a week or two before UPI05 created this bug. For the first few days after I thought it might be a WIP type of thing. After another week went by I wondered if it was just me.

Looks like it might be commit 50ed7b4e76dffb71a2fb970b2b1e60d846ea19f4 from pull 3574?

**Comment by airween** [comment]
> Looks like it might be commit [50ed7b4](https://github.com/coreruleset/coreruleset/commit/50ed7b4e76dffb71a2fb970b2b1e60d846ea19f4) from pull 3574?

Perhaps yes.

Conclusion: use the released v4.0.0. I can't suggest anything else at the moment.

**Comment by theseion** [comment]
Thanks for the analysis everyone.

We recently updated our regular expression generator. The generator is a Go program and uses `regexp/syntax` (RE2). In RE2, `\v` is simply the vertical tab character and it can start a range, just like any other character. In PCRE2, however, that is not the case. I've looked into the source code (https://github.com/PCRE2Project/pcre2/blob/3864abdb713f78831dd12d898ab31bbb0fa630b6/src/pcre2_compile.c#L3826):
```c
        /* Of the remaining escapes, only those that define characters are
        allowed in a class. None may start a range. */

        class_range_state = RANGE_NO;
        switch(escape)
          {
          <snip>

          case ESC_V:
          case ESC_v:
          *parsed_pattern++ = META_ESCAPE + escape;
          break;

         <snip>
        }
          <snip>

        /* Perl gives a warning unless a following hyphen is the last character
        in the class. PCRE throws an error. */

        if (ptr < ptrend - 1 && *ptr == CHAR_MINUS &&
            ptr[1] != CHAR_RIGHT_SQUARE_BRACKET)
          {
          errorcode = ERR50;
          goto FAILED;
          }
        }
```

`ERR50` is `invalid range in character class`. As the comment states, `\v` may not start a range expression (it doesn't say why, unfortunately).

I think the most portable approach to fixing this issue is to replace escapes like `\v` with their hexadecimal escape version (e.g., `\x0b`). This will work for PCRE, PCRE2 (because hexadecimal escape sequences can always start a range), and also RE2. I'll have a fix prepared in the coming days.

@dune73 we'll probably have to delay the next release until this has been fixed.

**Comment by theseion** [comment]
I've created a PR that should fix the issue. The regular expressions were generated with a fix to crs-toolchain for which I have created a draft PR (https://github.com/coreruleset/crs-toolchain/pull/137).

**Comment by dune73** [comment]
Duly noted. Thank you.
